### PR TITLE
Define _DEBUG for better ATL debugging

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Currently `ATLASSERT` macro won't work in GCC compilers, due to lack of `_DEBUG` macro.
Microsoft-style debugging needs `_DEBUG` macro definition.

JIRA issue: [CORE-17505](https://jira.reactos.org/browse/CORE-17505)

## Proposed changes

- Define `_DEBUG` macro as follows in some precompiled header that uses ATL module.

```c
#if !defined(_DEBUG) && !defined(NDEBUG)
    #define _DEBUG
#endif
```
- Fix `sdk/lib/atl/atlcom.h`.
- Fix `sdk/lib/atl/statreg.h`.

## TODO

- [x] Can build
- [x] Can test successful
- [ ] Get consensus of team

### CODE REVIEW

**Comment by SergeGautherie** [comment]
Could this define live in some ATL header(s), instead of being duplicated in every module than uses it?

**Comment by katahiromz** [comment]
Nope. Microsoft manual says that we have to define `_DEBUG` explicitly, so we follow their way.

File: sdk/lib/atl/statreg.h [hunk 1]
```diff
@@ -68,19 +68,19 @@ class CRegObject : public IRegistrarBase
 
     HRESULT STDMETHODCALLTYPE QueryInterface(const IID & /* riid */, void ** /* ppvObject */ )
     {
-        ATLASSERT(_T("statically linked in CRegObject is not a com object. Do not callthis function"));
+        ATLASSERT(FALSE && TEXT("statically linked in CRegObject is not a com object. Do not call this function"));
```
Comment (SergeGautherie) [inline sdk/lib/atl/statreg.h:71]: `call this`: fix the 2 others too.

**Comment by SergeGautherie** [comment]
> `#if !defined(_DEBUG) && !defined(NDEBUG)`

Why not check GCC/!MSVC, as you reported this is compiler dependent?
Especially, isn't your proposal wrong on MSVC?

> Microsoft manual says that we have to define `_DEBUG` explicitly, so we follow their way.

Which manual? What does it say exactly?

https://docs.microsoft.com/en-us/cpp/c-runtime-library/debug?view=msvc-160
`The compiler defines _DEBUG when you specify the /MTd or /MDd option. These options specify debug versions of the C run-time library.`
It looks like this define is meant to be checked, not to be set by code...
As in, maybe define it at cmake level, if that applies?

**Comment by katahiromz** [comment]
There are many NDEBUG-defined projects in ReactOS. So we can't define _DEBUG globally at CMake settings.

Thus, we have to define _DEBUG macro in specific projects individually. If so, we can either:

1. Define it at CMakeLists.txt of individual projects. Or:
2. Define it at precompiled (or top-level) header of individual projects.

**Comment by katahiromz** [comment]
Defining _DEBUG in CMakeLists.txt is better.
But I can't write the mechanism of CMake.

**Pull request closed without merging**